File: v8-core/src/benchmark/certificate.rs
//! V8 Evidence Dashboard & Policy Certificate (D-153).
//!
//! Enforces:
//! - Epistemic separation between Research Capability, Robustness, and Capital Projection.
//! - Multiplicative Readiness Index:
//!   Readiness = (Cap / 100) * EvidenceMultiplier * (Robustness / 100) * (Economic / 100) * 100
//! - Binary Robustness Seal & Hard Gate Vector verification.
//! - Multi-population evidence topology (12-month quad tape as single diagnostic cell).
//! - Terminal ASCII and HTML certificate rendering.

use serde::{Deserialize, Serialize};
use crate::benchmark::minerva::MinervaRobustness;
use crate::benchmark::projection::{CapitalOutcomeProjection, MonteCarloFutureResult};
use crate::benchmark::receipt::BenchmarkReceipt;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct PolicyCertificate {
    pub policy_id: String,
    pub receipt_id: String,
    pub status: String,
    pub research_capability_score: f64,
    pub evidence_multiplier: f64,
    pub minerva_robustness_score: f64,
    pub robustness_seal_status: String,
    pub economic_score: f64,
    pub readiness_index: f64,
    pub quad_tape_role: String,
    pub minerva: Option<MinervaRobustness>,
    pub monte_carlo: Option<MonteCarloFutureResult>,
}

impl PolicyCertificate {
    /// Generates a PolicyCertificate from evaluated benchmark artifacts.
    pub fn generate(
        receipt: &BenchmarkReceipt,
        projection: Option<&CapitalOutcomeProjection>,
    ) -> Self {
        let cap_score = (receipt.composite_capability_score * 100.0).clamp(0.0, 100.0);
        let evidence_multiplier = receipt.coverage_factor.clamp(0.10, 1.0);

        let minerva = receipt.minerva_robustness.clone();
        let (rob_score, seal_granted, seal_status) = match minerva.as_ref() {
            Some(m) => (m.effective_score, m.seal_granted, m.seal_status.clone()),
            None => (50.0, false, "SEAL_DENIED_NO_MINERVA_RUN".into()),
        };

        // Derive economic score from projection or receipt observations (0.0 .. 100.0)
        let economic_score = if let Some(proj) = projection {
            if let Some(ref mc) = proj.monte_carlo_futures {
                // Derived from median return, ruin penalty, and drawdown
                let ret_score = ((mc.p50_return_pct + 20.0) / 40.0).clamp(0.0, 1.0);
                let ruin_safety = (1.0 - (mc.risk_of_ruin_pct / 100.0)).clamp(0.0, 1.0);
                ((ret_score * 0.60 + ruin_safety * 0.40) * 100.0).round()
            } else if let Some(median_band) = proj.outcome_bands.iter().find(|b| (b.percentile - 0.50).abs() < 0.01) {
                ((median_band.return_bps / 500.0).clamp(0.0, 1.0) * 100.0).round()
            } else {
                60.0
            }
        } else {
            60.0
        };

        // Multiplicative Readiness Index:
        // Readiness = (Cap / 100) * EvidenceMultiplier * (Rob / 100) * (Econ / 100) * 100
        let readiness = (cap_score / 100.0)
            * evidence_multiplier
            * (rob_score / 100.0)
            * (economic_score / 100.0)
            * 100.0;

        let is_production_approved = readiness >= 80.0
            && seal_granted
            && receipt.projection_grade.allows_forward_probability()
            && receipt.gate_vector.all_passed();

        let status = if is_production_approved {
            "STATUS: Production Ready".to_string()
        } else {
            "STATUS: Research Candidate NOT Production Approved".to_string()
        };

        let quad_tape_role = "HISTORICAL DIAGNOSTIC CELL (Non-universal evaluation fold)".to_string();

        let monte_carlo = projection.and_then(|p| p.monte_carlo_futures.clone());

        Self {
            policy_id: receipt.policy_id.clone(),
            receipt_id: receipt.receipt_id.clone(),
            status,
            research_capability_score: cap_score,
            evidence_multiplier,
            minerva_robustness_score: rob_score,
            robustness_seal_status: seal_status,
            economic_score,
            readiness_index: (readiness * 10.0).round() / 10.0,
            quad_tape_role,
            minerva,
            monte_carlo,
        }
    }

    /// Renders clean terminal ASCII certificate matching user specifications.
    pub fn render_ascii(&self) -> String {
        let bar = |val: f64, max: f64| -> String {
            let pct = (val / max).clamp(0.0, 1.0);
            let filled = (pct * 30.0).round() as usize;
            let empty = 30 - filled;
            format!("[{}{}]", "|".repeat(filled), ".".repeat(empty))
        };

        let mut out = String::new();
        out.push_str("======================================================================\n");
        out.push_str("               V8 EVIDENCE DASHBOARD & POLICY CERTIFICATE             \n");
        out.push_str("======================================================================\n");
        out.push_str(&format!("Policy Target:  {}\n", self.policy_id));
        out.push_str(&format!("Receipt Digest: {}\n", self.receipt_id));
        out.push_str(&format!("Final Verdict:  {}\n", self.status));
        out.push_str("----------------------------------------------------------------------\n");

        out.push_str("1. RESEARCH CAPABILITY SCORE (Infrastructure & Integrity):\n");
        out.push_str(&format!("   Score: {:>5.1} / 100  {}\n", self.research_capability_score, bar(self.research_capability_score, 100.0)));
        out.push_str(&format!("   Evidence Multiplier: {:.2}\n", self.evidence_multiplier));
        out.push_str("----------------------------------------------------------------------\n");

        out.push_str("2. ECONOMIC EVIDENCE & MINERVA ROBUSTNESS (arXiv:2608.23808):\n");
        out.push_str(&format!("   Minerva Score:  {:>5.1} / 100  {}\n", self.minerva_robustness_score, bar(self.minerva_robustness_score, 100.0)));
        out.push_str(&format!("   Robustness Seal: {}\n", self.robustness_seal_status));
        
        if let Some(ref m) = self.minerva {
            out.push_str("   Validation Gates:\n");
            out.push_str(&format!("     DSR Gate:    [{:>4}] (Margin: {:+.3})\n", m.gate_vector.dsr_gate.as_str(), m.margins.dsr_margin));
            out.push_str(&format!("     PBO Gate:    [{:>4}] (Margin: {:+.3})\n", m.gate_vector.pbo_gate.as_str(), m.margins.pbo_margin));
            out.push_str(&format!("     SPA Gate:    [{:>4}] (Margin: {:+.3})\n", m.gate_vector.spa_gate.as_str(), m.margins.spa_margin));
            out.push_str(&format!("     MinTRL Gate: [{:>4}] (Margin: {:+.1} days)\n", m.gate_vector.min_trl_gate.as_str(), m.margins.min_trl_margin));
            out.push_str(&format!("     Regime Gate: [{:>4}] (Margin: {:+.1} bps)\n", m.gate_vector.regime_stability_gate.as_str(), m.margins.regime_stability_margin));
        }

        out.push_str(&format!("   Evidence Topology: {}\n", self.quad_tape_role));
        out.push_str("----------------------------------------------------------------------\n");

        out.push_str("3. RISK-ADJUSTED CAPITAL PROJECTION ($1,000 Initial, 1-Year Horizon):\n");
        if let Some(ref mc) = self.monte_carlo {
            out.push_str(&format!("   10,000 Monte Carlo Futures (Horizon: {} trades):\n", mc.horizon_trades));
            out.push_str(&format!("     P5:   ${:>7.2}  ({:>+6.1}%)\n", mc.p5_terminal_usd, mc.p5_return_pct));
            out.push_str(&format!("     P25:  ${:>7.2}  ({:>+6.1}%)\n", mc.p25_terminal_usd, mc.p25_return_pct));
            out.push_str(&format!("     P50:  ${:>7.2}  ({:>+6.1}%) [Median]\n", mc.p50_terminal_usd, mc.p50_return_pct));
            out.push_str(&format!("     P75:  ${:>7.2}  ({:>+6.1}%)\n", mc.p75_terminal_usd, mc.p75_return_pct));
            out.push_str(&format!("     P95:  ${:>7.2}  ({:>+6.1}%)\n", mc.p95_terminal_usd, mc.p95_return_pct));
            let worst_usd = mc.initial_capital_usd * (1.0 + mc.worst_scenario_return_pct / 100.0);
            let best_usd = mc.initial_capital_usd * (1.0 + mc.best_scenario_return_pct / 100.0);
            out.push_str(&format!("     Worst: ${:>7.2} ({:>+6.1}%)  Best: ${:>7.2} ({:>+6.1}%)\n", worst_usd, mc.worst_scenario_return_pct, best_usd, mc.best_scenario_return_pct));
            out.push_str(&format!("     Risk of Ruin (Drawdown >= 30%): {:.1}%\n", mc.risk_of_ruin_pct));
            out.push_str(&format!("   Notice: {}\n", mc.conditional_notice));
        } else {
            out.push_str("   [Underpowered sample or diagnostic run: extreme percentiles suppressed]\n");
        }

        out.push_str("----------------------------------------------------------------------\n");
        out.push_str(&format!("READINESS INDEX: {:>5.1} / 100\n", self.readiness_index));
        out.push_str(&format!(
            "Formula: Cap ({:.1}) * Evidence ({:.2}) * Robustness ({:.1}) * Economic ({:.1}) / 100^2\n",
            self.research_capability_score, self.evidence_multiplier, self.minerva_robustness_score, self.economic_score
        ));
        out.push_str("======================================================================\n");

        out
    }
}

File: v8-next/src/v8_next/evaluation/certificate.py
"""V8 Evidence Dashboard & Policy Certificate (D-153, D-152 §5).

Enforces:
- Epistemic separation between Research Capability, Robustness, and Capital Projection.
- Multiplicative Readiness Index:
  Readiness = (Cap / 100) * EvidenceMultiplier * (Robustness / 100) * (Economic / 100) * 100
- Binary Robustness Seal & Hard Gate Vector verification.
- Multi-population evidence topology (12-month quad tape as single diagnostic cell).
- Terminal ASCII and HTML certificate rendering.
"""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, ConfigDict

from v8_next.evaluation.benchmark_receipt import (
    BenchmarkReceipt,
    GateState,
    ReadinessStatus,
)


def _render(value: float | None, digits: int = 1) -> str:
    """A missing measurement renders as MISSING, never as a zero."""
    return "MISSING" if value is None else f"{value:.{digits}f}"


class PolicyCertificate(BaseModel):
    """Canonical D-153 Policy Certificate and Evidence Dashboard."""

    model_config = ConfigDict(frozen=True)

    policy_id: str
    receipt_id: str
    status: str
    authority_verdict: str
    gates: list[tuple[str, str]]
    #: ``None`` wherever the measurement does not exist (NX08.R2). The previous
    #: defaults (robustness 50.0, economic 60.0) were fabricated inputs to the
    #: readiness formula and are gone: a missing factor now makes the readiness
    #: index missing instead of quietly producing a number.
    research_capability_score: float | None
    evidence_multiplier: float | None
    minerva_robustness_score: float | None
    robustness_seal_status: str
    economic_score: float | None
    readiness_index: float | None
    quad_tape_role: str
    derivation: dict[str, Any] | None = None
    missing_measurements: tuple[str, ...] = ()
    readiness_upper_bound: float | None = None
    minerva: Any | None = None
    monte_carlo: Any | None = None

    @classmethod
    def generate(
        cls,
        receipt: BenchmarkReceipt,
        projection: Any | None = None,
        minerva: Any | None = None,
    ) -> PolicyCertificate:
        """Generates a PolicyCertificate from evaluated benchmark artifacts."""
        raw_cap = receipt.capability_score
        cap_score = (
            None if raw_cap is None else round(min(100.0, max(0.0, raw_cap)), 1)
        )
        raw_coverage = receipt.coverage_factor
        evidence_multiplier = (
            None
            if raw_coverage is None
            else round(min(1.0, max(0.10, raw_coverage)), 2)
        )

        # Robustness: no default. Without a Minerva run there is no robustness
        # measurement, and the seal says so.
        seal_status = "SEAL_DENIED_NO_MINERVA_RUN"
        rob_score: float | None = None
        if minerva is not None:
            rob_score = getattr(minerva, "effective_score", None)
            seal_status = getattr(minerva, "seal_status", "SEAL_DENIED_NO_MINERVA_RUN")

        # Economic: no default in a diagnostic cell. A missing projection stays
        # missing instead of being scored 60.0.
        economic_score: float | None = None
        if projection is not None:
            economic_score = getattr(projection, "economic_score", None)

        # Multiplicative Readiness Index, evaluated only over factors that exist.
        factors: dict[str, float | None] = {
            "research_capability_score": cap_score,
            "evidence_multiplier": evidence_multiplier,
            "minerva_robustness_score": rob_score,
            "economic_score": economic_score,
        }
        missing = tuple(name for name, value in factors.items() if value is None)
        readiness_index: float | None = None
        readiness_upper_bound: float | None = None
        if not missing:
            # narrowed by "not missing": every factor exists, so the formula is
            # evaluated over real measurements only
            assert cap_score is not None
            assert evidence_multiplier is not None
            assert rob_score is not None
            assert economic_score is not None
            readiness_index = round(
                (cap_score / 100.0)  # type: ignore[operator]
                * evidence_multiplier  # type: ignore[operator]
                * (rob_score / 100.0)  # type: ignore[operator]
                * (economic_score / 100.0)  # type: ignore[operator]
                * 100.0,
                1,
            )
            # upper bound derived from the same formula: capability at its ceiling
            readiness_upper_bound = round(
                1.0
                * evidence_multiplier  # type: ignore[operator]
                * (rob_score / 100.0)  # type: ignore[operator]
                * (economic_score / 100.0)  # type: ignore[operator]
                * 100.0,
                1,
            )
        derivation = {
            "formula": "capability * evidence_multiplier * robustness * economic / 100**2",
            "transform_version": "readiness-v2-missing-aware",
            "raw_measurements": {
                "capability_score": receipt.capability_score,
                "coverage_factor": receipt.coverage_factor,
                "minerva_effective_score": rob_score,
                "projection_economic_score": economic_score,
            },
            "denominator": 100.0**2,
            "weights": {name: 1.0 for name in factors},
            "binding": {
                "receipt_digest": receipt.receipt_digest,
                "digest_version": receipt.digest_version,
                "gate_vector": {
                    field: getattr(receipt.gates, field).value
                    for field in sorted(receipt.gates.__class__.model_fields)
                },
            },
            "missing_measurements": list(missing),
        }

        verdict = receipt.gates.readiness()
        if verdict.status == ReadinessStatus.HardFailure:
            status = "STATUS: BLOCKED (hard gate failure)"
            authority_verdict = "VERDICT: BLOCKED (hard gate failure)"
        elif verdict.status == ReadinessStatus.Certified:
            status = "STATUS: Research Candidate Ready For Review NOT Production Approved"
            if receipt.gates.g9_live_realization == GateState.PASS:
                if receipt.gates.g8_prospective_shadow == GateState.PASS:
                    authority_verdict = "VERDICT: SUPPORTED_EDGE (Live venue fills settled & StatutoryClaimRecord registered)"
                else:
                    authority_verdict = "VERDICT: READY_NOT_CLAIMED (StatutoryClaimRecord registered; human review required for live capital)"
            else:
                authority_verdict = "VERDICT: NO_ECONOMIC_CLAIM (diagnostic instrument only; no economic authority may be derived)"
        else:
            status = "STATUS: NO_ECONOMIC_CLAIM (Research Candidate, NOT Production Approved)"
            authority_verdict = "VERDICT: NO_ECONOMIC_CLAIM (diagnostic instrument only; no economic authority may be derived)"

        quad_tape_role = "HISTORICAL DIAGNOSTIC CELL (Non-universal evaluation fold)"

        gates = [
            (
                f"{ev.descriptor.canonical_id}::{ev.descriptor.vector_field}",
                ev.state.value,
            )
            for ev in verdict.evaluations
        ]

        return cls(
            policy_id=receipt.policy_id,
            receipt_id=receipt.receipt_digest,
            status=status,
            authority_verdict=authority_verdict,
            gates=gates,
            research_capability_score=cap_score,
            evidence_multiplier=evidence_multiplier,
            minerva_robustness_score=rob_score,
            robustness_seal_status=seal_status,
            economic_score=economic_score,
            readiness_index=readiness_index,
            quad_tape_role=quad_tape_role,
            derivation=derivation,
            missing_measurements=missing,
            readiness_upper_bound=readiness_upper_bound,
            minerva=minerva,
            monte_carlo=None,
        )

    def render_ascii(self) -> str:
        """Renders clean terminal ASCII certificate matching D-153 Rust reference."""

        def bar(val: float, max_val: float) -> str:
            pct = min(1.0, max(0.0, val / max_val))
            filled = int(round(pct * 30.0))
            empty = 30 - filled
            return f"[{'|' * filled}{'.' * empty}]"

        lines = [
            "======================================================================",
            "               V8 EVIDENCE DASHBOARD & POLICY CERTIFICATE             ",
            "======================================================================",
            f"Policy Target:  {self.policy_id}",
            f"Receipt Digest: {self.receipt_id}",
            f"Final Verdict:  {self.status}",
            f"Authority:      {self.authority_verdict}",
            "----------------------------------------------------------------------",
            "0. HARD GATE VECTOR G0-G9 (non-compensable, D-152 §5):",
        ]
        for gate_name, state in self.gates:
            lines.append(f"   {gate_name:<58} [{state}]")
        lines.extend(
            [
                "----------------------------------------------------------------------",
                "1. RESEARCH CAPABILITY SCORE (Infrastructure & Integrity):",
                f"   Score: {_render(self.research_capability_score)} / 100  {bar(self.research_capability_score or 0.0, 100.0)}",
                f"   Evidence Multiplier: {_render(self.evidence_multiplier, 2)}",
                "----------------------------------------------------------------------",
                "2. ECONOMIC EVIDENCE & MINERVA ROBUSTNESS (arXiv:2608.23808):",
                f"   Minerva Score:  {_render(self.minerva_robustness_score)} / 100  {bar(self.minerva_robustness_score or 0.0, 100.0)}",
                f"   Robustness Seal: {self.robustness_seal_status}",
                f"   Evidence Topology: {self.quad_tape_role}",
                "----------------------------------------------------------------------",
                "3. RISK-ADJUSTED CAPITAL PROJECTION ($1,000 Initial, 1-Year Horizon):",
                "   [Underpowered sample or diagnostic run: extreme percentiles suppressed]",
                "----------------------------------------------------------------------",
                f"READINESS INDEX: {_render(self.readiness_index)} / 100",
                "Formula: Cap ({cap}) * Evidence ({ev}) * Robustness ({rob}) * Economic ({eco}) / 100^2".format(
                    cap=_render(self.research_capability_score),
                    ev=_render(self.evidence_multiplier, 2),
                    rob=_render(self.minerva_robustness_score),
                    eco=_render(self.economic_score),
                ),
                f"Missing measurements: {', '.join(self.missing_measurements) if self.missing_measurements else 'none'}",
                "======================================================================",
                "",
            ]
        )
        return "\n".join(lines)
